Translate Chapter 10 to PTBR (#18)
JuliaIntro/JuliaIntroBR.jl: book/src/chap10.asciidoc

[[chap10]]
== Listas

Este capítulo apresenta um dos tipos internos mais úteis do Julia, listas. Você também aprenderá sobre objetos e o que pode acontecer quando você tiver mais de um nome para o mesmo objeto.

=== Uma Lista é uma Sequência

> Como uma string, ua _lista_ é uma sequência de valores. Em uma string, os valores são caracteres; em um vetor, pode ser de qualquer tipo. Os valores de um vetor são chamados de _elementos_ ou por vezes _itens_.

Reviewer: ua -> uma

> Como uma string, ua _lista_ é uma sequência de valores. Em uma string, os valores são caracteres; em um vetor, pode ser de qualquer tipo. Os valores de um vetor são chamados de _elementos_ ou por vezes _itens_.

Reviewer: "em um vetor, pode ser de qualquer tipo .." -> "em um vetor, eles podem ser de qualquer tipo .."

(((lista)))(((element)))(((item)))

Existem diversas maneiras de criar uma nova lista; a mais simples é inserir os elementos entre colchetes (+[ ]+):
(((colchetes)))

[source,julia]
----
[10, 20, 30, 40]
["crunchy frog", "ram bladder", "lark vomit"]
----

O primeiro exemplo é uma lista de quatro inteiros. O segundo é uma lista de três strings. Os elementos de uma lista não precisam ser do mesmo tipo. A seguinte lista contém uma string, um float, um inteiro, e uma outra lista:

[source,julia]
----
["bruxa", 2.0, 5, [10, 20]]
----

Uma lista com outra lista é _aninhada_.
(((aninhado)))

Uma lista que não contém elementos é chamado de lista vazia; você pode criar uma lista vazia com colchetes vazios, +[]+.

> (((vetor vazio)))

Reviewer: "vetor vazio" -> "lista vazia"

Como esperado, podemos atribuir valores de listas a variáveis:

[source,@julia-repl-test chap10]
----
julia> vegetais = ["Cenoura", "Brócolis", "Alface"];

julia> números = [42, 123];

julia> vazia = [];

julia> print(vegetais, " ", números, " ", vazio)
["Cenoura", "Brócolis", "Alface"] [42, 123] Any[]
----

A função +typeof+ pode ser usada para encontrar o tipo de lista:
(((typeof)))

[source,@julia-repl-test chap10]
----
julia> typeof(vegetais)
Array{String,1}
julia> typeof(números)
Array{Int64,1}
julia> typeof(vazia)
Array{Any,1}
----

> O tipo de lista é especificado entre as chaves e é composto pelo tipo e um número. O número indica a dimensão. A lista +vazia+ contém valores do tipo +Any+., isto é, pode ser designado a valores de todos os tipos.

Reviewer: "o número indica a dimensão" -> "o número indica as dimensões"

> (((Listas)))((("tipo", "Base", "Vwtor", see="Array")))(((Any)))((("type", "Base", "Any", see="Any")))

Reviewer: "Vwtor" -> "Lista" ou "Vetor"
" .. , see="Array" " -> ".. , see="Lista" "

=== Listas são mutáveis

A sintaxe para acessar elementos de uma lista é a mesma para acessar caracteres de uma string - colchetes. A expressão dentro dos colchetes especifica o índice. Lembre-se que o índice começa em 1:
(((colchetes)))(((índice)))

[source,@julia-repl-test chap10]
----
julia> vegetais[1]
"Cenoura"
----

> Assim como strings, listas são _mutáveis_. Quando os colchetes aparecem no lado esquerdo de uma atribuição, o elemento da lista que está designado é identificado:

Reviewer: "Assim como strings .. " -> "Ao contrário de strings .."
Aqui vale a pena mudar "Quando os colchetes aparecem no lado esquerdo de uma atribuição, o elemento da lista que está designado é identificado: " -> "Quando os colchetes aparecem no lado esquerdo de uma atribuição, eles identificam o elemento da lista que será atribuido:"

(((mutável)))(((declaração de atribuição)))

[source,@julia-repl-test chap10]
----
julia> números[2] = 5
5
julia> print(números)
[42, 5]
----

O segundo elemento de +números+, que costumava ser 1997, agora é 7.

<<fig10-1>> mostra o diagrama de pilha para "vegetais", "números" e "vazio".
(((diagrama de pilha)))

[[fig10-1]]
.Diagrama de pilha
image::images/fig101.svg[]

> Vetores são representados por caixas e os elementos dos vetores dentro. +vegetais+ refere-se a uma lista com três elementos indexados +1+, +2+ e +3+. +números+ contém dois elementos; o diagrama mostra que a lista do segundo elemento foi reatribuída de +123+ para +5+. +vazio+ refere-se a uma lista sem elementos.

Reviewer: "Vetores" -> "Listas"

Os índices das listas funcionam do mesmo jeito que os índices de string (mas sem as ressalvas do UTF-8):

* Qualquer expressão inteira pode ser utilizada como um índice.

> * Se você tentar ler ou escrever um elemento que não existe, será exibido +BoundsError+

Reviewer: "será exibido" -> "você receberá um"

[[chap10]]
== Listas

> Este capítulo apresenta um dos tipos internos mais úteis do Julia, as listas. Você também aprenderá sobre objetos e o que pode acontecer quando você tiver mais de um nome para o mesmo objeto.

Reviewer: ```suggestion
Este capítulo apresenta um dos tipos internos mais úteis do Julia, as listas. Você também aprenderá sobre objetos e o que pode acontecer quando se tem mais de um nome para o mesmo objeto.
```
Uma sugestão para evitar a repetição do pronome

=== Uma Lista é uma Sequência

> Como uma string, uma _lista_ é uma sequência de valores. Em uma string, os valores são caracteres; em um lista, pode ser de qualquer tipo. Os valores de um lista são chamados de _elementos_ ou por vezes _itens_.

Reviewer: ```suggestion
Assim como uma string, uma _lista_ é uma sequência de valores. Em uma string, os valores são caracteres; em uma lista, eles podem ser de qualquer tipo. Os valores de uma lista são chamados de _elementos_ ou por vezes _itens_.
```

(((lista)))(((elemento)))(((item)))

> Existem diversas maneiras de criar uma nova lista; a mais simples é inserir os elementos entre colchetes (+[ ]+), exemplo:

Reviewer: ```suggestion
Existem diversas maneiras de criar uma nova lista; a mais simples é inserir os elementos entre colchetes (+[ ]+), da seguinte forma:
```

> (((colchetes)))

Reviewer: ```suggestion
(((operador colchetes)))
```

[source,julia]
----
[10, 20, 30, 40]
["sapo", "coelho", "aranha"]
----

O primeiro exemplo é uma lista de quatro inteiros. O segundo é uma lista de três strings. Os elementos de uma lista não precisam ser do mesmo tipo. A seguinte lista contém uma string, um float, um inteiro e uma outra lista:

[source,julia]
----
["bruxa", 2.0, 5, [10, 20]]
----

> Uma lista com outra lista é dita _aninhada_.

Reviewer: ```suggestion
Uma lista dentro de outra lista é dita _aninhada_.
```

(((aninhada)))

Uma lista que não contém elementos é chamada de lista vazia; você pode criar uma lista vazia com colchetes vazios, +[]+.
(((lista vazia)))

Como esperado, podemos atribuir valores de listas a variáveis:

[source,@julia-repl-test chap10]
----
julia> vegetais = ["Cenoura", "Brócolis", "Alface"];

> julia> números = [42, 1997];

Reviewer: ```suggestion
julia> números = [42, 123];
```

julia> vazia = [];

julia> print(vegetais, " ", números, " ", vazio)
["Cenoura", "Brócolis", "Alface"] [42, 123] Any[]
----

A função +typeof+ pode ser usada para encontrar o tipo de lista:
(((typeof)))

[source,@julia-repl-test chap10]
----
julia> typeof(vegetais)
Array{String,1}
julia> typeof(números)
Array{Int64,1}
julia> typeof(vazia)
Array{Any,1}
----

> O tipo de lista é especificado entre as chaves e é composto pelo tipo e um número. O número indica as dimensões. A lista +vazia+ contém valores do tipo +Any+., isto é, pode ser designado a valores de todos os tipos.

Reviewer: ```suggestion
O tipo de lista é especificado entre as chaves e é composto por um tipo e um número. O número indica as dimensões. A lista +vazia+ contém valores do tipo +Any+., isto é, ela pode conter valores de todos os tipos.
```

(((Lista)))((("tipo", "Base", "Lista", see="Lista")))(((Any)))((("tipo", "Base", "Any", see="Any")))

=== Listas são Mutáveis

> A sintaxe para acessar elementos de uma lista é a mesma para acessar caracteres de uma string - colchetes. A expressão dentro dos colchetes especifica o índice. Lembre-se que o índice começa em 1:

Reviewer: ```suggestion
A sintaxe para acessar elementos de uma lista é a mesma para acessar caracteres de uma string com o operador colchetes. A expressão dentro dos colchetes especifica o índice. Lembre-se que o índice começa em 1:
```

> (((colchetes)))(((índice)))

Reviewer: ```suggestion
(((operador colchetes)))(((índice)))
```

[source,@julia-repl-test chap10]
----
julia> vegetais[1]
"Cenoura"
----

Ao contrário de strings, listas são _mutáveis_. Quando os colchetes aparecem no lado esquerdo de uma atribuição, eles identificam o elemento da lista que será atribuido:
(((mutável)))(((comando de atribuição)))

[source,@julia-repl-test chap10]
----

> julia> números[2] = 7

Reviewer: ```suggestion
julia> números[2] = 5
```

Reviewer: mudar pra não ter conflito com a imagem

> 7

Reviewer: ```suggestion
5
```

julia> print(números)

> [42, 7]

Reviewer: ```suggestion
[42, 5]
```

----

> O segundo elemento de +números+, que costumava ser 1997, agora é 7.

Reviewer: ```suggestion
O segundo elemento de +números+, que costumava ser 123, agora é 7.
``

> O segundo elemento de +números+, que costumava ser 123, agora é 7.

Reviewer: ```suggestion
O segundo elemento de +números+, que costumava ser 123, agora é 5.
```

<<fig10-1>> mostra o diagrama de pilha para "vegetais", "números" e "vazia".

> (((diagrama de pilha)))

Reviewer: ```suggestion
(((diagrama de estado)))
```

> <<fig10-1>> mostra o diagrama de estado para "vegetais", "números" e "vazia".

Reviewer: ```suggestion
<<fig10-1>> mostra o diagrama de estado para +vegetais+, +números+ e +vazia+.
```

Reviewer: nomes de variáveis.

[[fig10-1]]

> .Diagrama de pilha

Reviewer: ```suggestion
.Diagrama de estado
```

image::images/fig101.svg[]

> Listas são representadas por caixas e pelos elementos das listas dentro. +vegetais+ refere-se a uma lista com três elementos indexados +1+, +2+ e +3+. +números+ contém dois elementos; o diagrama mostra que a lista do segundo elemento foi reatribuída de +1997+ para +7+. +vazia+ refere-se a uma lista sem elementos.

Reviewer: ```suggestion
Listas são representadas por caixas e pelos elementos das listas dentro. +vegetais+ refere-se a uma lista com três elementos indexados +1+, +2+ e +3+. +números+ contém dois elementos; o diagrama mostra que o valor do segundo elemento foi reatribuído de +123+ para +7+. +vazia+ refere-se a uma lista sem elementos.
```

> Os índices das listas funcionam do mesmo jeito que os índices de string (mas sem as ressalvas do UTF-8):

Reviewer: ```suggestion
Os índices das listas funcionam do mesmo jeito que os índices de strings (mas sem as ressalvas do UTF-8):
```

* Qualquer expressão inteira pode ser utilizada como um índice.

> * Se você tentar ler ou escrever um elemento que não existe, você obterá um +BoundsError+.

Reviewer: ```suggestion
* Se você tentar ler ou escrever um elemento que não existe, um +BoundsError+ (erro de limite) será gerado.
```

> * Se você tentar ler ou escrever um elemento que não existe, um +BoundsError+ (erro de limite) será gerado.

Reviewer 2: ```suggestion
* Se você tentar ler ou escrever um elemento que não existe, um +BoundsError+ (erro de limites) será gerado.
```

* A palavra-chave +end+ aponta para o último índice da lista.
(((end)))

O operador +∈+ também funciona em listas:
(((em)))

[source,@julia-repl-test chap10]
----
julia> "Cenoura" ∈ vegetais
true
julia> "Beterraba" in vegetais
false
----

=== Percorrendo uma Lista

A maneira mais comum de percorrer os elementos de uma lista é através do laço +for+. A sintaxe é a mesma para strings:

> (((percorrer)))(((comando for)))

Reviewer: ```suggestion
(((percorrer)))(((declaração for)))
```

[source,@julia-setup chap10]
----
for vegetal in vegetais
    println(vegetal)
end
----

Isso funciona bem se você precisa apenas ler os elementos de uma lista. Mas se você deseja escrever ou atualizar os elementos, você precisa dos índices. Um jeito comum de fazer isto é utilizando a função embutida +eachindex+:
(((eachindex)))((("função", "Base", "eachindex", see="eachindex")))

[source,@julia-setup chap10]
----
for i in eachindex(números)
    números[i] = números[i] * 2
end
----

> Esse laço percorre a lista e atualiza cada elemento. +length+ retorna o número de elementos de uma lista. Toda vez que o laço é percorrido, +i+ obtém o índice do próximo elemento. O comando de atribuição no corpo usa +i+ para ler os valores antigos do elemento e designa ao novo valor:

Reviewer: ```suggestion
Esse laço percorre a lista e atualiza cada elemento. +length+ retorna o número de elementos de uma lista. Toda vez que o laço é percorrido, +i+ obtém o índice do próximo elemento. O comando de atribuição no corpo usa +i+ para ler o valor antigo do elemento e atribuir ao novo valor:
```

[source,@julia-repl-test chap10]
----
julia> typeof(vegetais)
Array{String,1}
julia> typeof(números)
Array{Int64,1}
julia> typeof(vazia)
Array{Any,1}
----

O tipo de lista é especificado entre as chaves e é composto por um tipo e um número. O número indica as dimensões. A lista +vazia+ contém valores do tipo +Any+., isto é, ela pode conter valores de todos os tipos.
(((Lista)))((("tipo", "Base", "Lista", see="Lista")))(((Any)))((("tipo", "Base", "Any", see="Any")))

=== Listas são Mutáveis

A sintaxe para acessar elementos de uma lista é a mesma para acessar caracteres de uma string com o operador colchetes. A expressão dentro dos colchetes especifica o índice. Lembre-se que o índice começa em 1:
(((operador colchetes)))(((índice)))

[source,@julia-repl-test chap10]
----
julia> vegetais[1]
"Cenoura"
----

Ao contrário de strings, listas são _mutáveis_. Quando os colchetes aparecem no lado esquerdo de uma atribuição, eles identificam o elemento da lista que será atribuido:
(((mutável)))(((comando de atribuição)))

[source,@julia-repl-test chap10]
----
julia> números[2] = 7
7
julia> print(números)
[42, 7]
----

O segundo elemento de +números+, que costumava ser 123, agora é 7.

> ``

Reviewer: ```suggestion

```

> <<fig10-1>> mostra o diagrama de pilha para "vegetais", "números" e "vazia".

Reviewer: ```suggestion
<<fig10-1>> mostra o diagrama de pilha para +vegetais+, +números+ e +vazia+.
```

(((diagrama de estado)))

[[fig10-1]]
.Diagrama de estado
image::images/fig101.svg[]

> Listas são representadas por caixas e pelos elementos das listas dentro. +vegetais+ refere-se a uma lista com três elementos indexados +1+, +2+ e +3+. +números+ contém dois elementos; o diagrama mostra que o valor do segundo elemento foi reatribuído de +123+ para +7+. +vazia+ refere-se a uma lista sem elementos.

Reviewer: ```suggestion
Uma lista é representada por uma caixa e pelos elementos da lista dentro dela. +vegetais+ refere-se a uma lista com três elementos indexados +1+, +2+ e +3+. +números+ contém dois elementos; o diagrama mostra que o valor do segundo elemento foi reatribuído de +123+ para +7+. +vazia+ refere-se a uma lista sem elementos.
```

Os índices das listas funcionam do mesmo jeito que os índices de strings (mas sem as ressalvas do UTF-8):

* Qualquer expressão inteira pode ser utilizada como um índice.

* Se você tentar ler ou escrever um elemento que não existe, um +BoundsError+ (erro de limite) será gerado.

* A palavra-chave +end+ aponta para o último índice da lista.
(((end)))

O operador +∈+ também funciona em listas:
(((em)))

[source,@julia-repl-test chap10]
----
julia> "Cenoura" ∈ vegetais
true
julia> "Beterraba" in vegetais
false
----

=== Percorrendo uma Lista

> A maneira mais comum de percorrer os elementos de uma lista é através do laço +for+. A sintaxe é a mesma para strings:

Reviewer: ```suggestion
A maneira mais comum de percorrer os elementos de uma lista é através de um laço +for+. A sintaxe é a mesma para strings:
```

----

> === Atravessando uma Lista

Reviewer: "Atravessando" -> "Percorrendo"

O jeito mais comum de atraversar os elementos de uma lista é através do laço +for+. A sintaxe é a mesma para strings:
(((travessia)))(((atribuição for)))

[source,@julia-setup chap10]
----
for queijo in queijos
    println(queijo)
end
----

> Isso funciona bem se você for apenas ler os alementos de uma lista. Mas se você precisa escrever ou atualizar os elementos, você precisa dos índices. Um jeito comum de fazer isso é utilizando a função embutida +eachindex+:

Reviewer: "... se você for apenas ler os elementos de uma lista .." -> " ... se você precisa somente ler os elementos de uma lista .. "

(((eachindex)))((("função", "Base", "eachindex", see="eachindex")))

[source,@julia-setup chap10]
----
for i in eachindex(números)
    números[i] = números[i] * 2
end
----

> Esse laço atravessa a lista e atualiza cada elemento. +lenght+ retorna o número de elementos de uma lista. Cada vez no laço +i+ obtém o índice do próximo elemento. A declaração de atribuição no corpo usa +i+ para ler os valores antigos do elemento e desgina ao novo valor:

Reviewer: "atravessa" -> "percorre"
"+lenght+" -> "+length+"
"desgina" -> "designa"

> Esse laço atravessa a lista e atualiza cada elemento. +lenght+ retorna o número de elementos de uma lista. Cada vez no laço +i+ obtém o índice do próximo elemento. A declaração de atribuição no corpo usa +i+ para ler os valores antigos do elemento e desgina ao novo valor:

Reviewer: "Cada vez no laço +i+ obtém o índice ..." -> "Toda vez que o laço é percorrido +i+ obtém o indíce .. "

> O laço +for+ sobre uma lista vazia nunca executa o corpo:

Reviewer: "O laço +for+ sobre uma lista .. " -> "Utilizar um laço +for+ sobre uma lista .. "

['c', 'd', 'e', 'f']
----

O Operador de Fatia +[:]+, faz uma cópia em toda a lista:
(((cópia)))

[source,@julia-repl-test chap10]
----
julia> print(t[:])
['a', 'b', 'c', 'd', 'e', 'f']
----

Como as listas são mutáveis, geralmente é útil fazer uma cópia antes de executar operações que modificam listas.

> O Operador de Fatais no lado esquerdo de uma atribuição pode atualizar múltiplos elementos:

Reviewer: "Fatais" -> "Fatias"

['c', 'd', 'e', 'f']
----

> O Operador de Fatia +[:]+, faz uma cópia em toda a lista:

Reviewer: ```suggestion
O Operador de Fatia +[:]+, faz uma cópia de toda a lista:
```

----

=== Biblioteca de Listas

> Julia fornece função que operam com listas. Por exemplo, +push!+ adiciona um novo elemento no final de uma lista:

Reviewer: "função " -> "funções"
" .. um novo elemento no final .. " -> " .. um novo elemento ao final .. "

['a', 'b', 'c', 'd']
----

> +append!+ adiciona elementos da segunda lista ao final da primeira:

Reviewer: ```suggestion
+append!+ adiciona os elementos da segunda lista ao final da primeira:
```

['a', 'b', 'c', 'd', 'e']
----

> Esse exemplo deixa +t2+ sem modificação.

Reviewer: "modificação" -> "modificações"

+sort!+ organiza os elementos da lista do menor para o maior:

> (((sor!)))((("função", "Base", "sort!", see="sort!")))

Reviewer: "sor!" -> "sort!"

['a', 'b', 'c', 'd', 'e']
----

> +sort_ retorna uma cópia dos elementos da lista em ordem:

Reviewer: "+sort_" -> "+sort+"

> +sort!+ retorna uma cópia dos elementos da lista em ordem:

Reviewer: ```suggestion
+sort+ retorna uma cópia dos elementos da lista em ordem:
```

> (((sort)))((("função", "Base", "sort"m, see="sort")))

Reviewer: ```suggestion
(((sort)))((("função", "Base", "sort", see="sort")))
```

[NOTE]
====

> Como uma convenção de estilo em Julia, +! + É anexado a nomes de funções que modificam seus argumentos.

Reviewer: tirar o espaço de +! + e maiúscula de "É"

(((!)))
====

=== Mapeamento, Filtro e Redução

> Para somar todos os números de uma lista, podemos utilizar um laço como o seguinte:

Reviewer: ```suggestion
Para somar todos os números de uma lista, podemos utilizar um laço da seguinte forma:
```

> Para aumentar todos os números de uma lista, podemos utilizar um laço como o seguinte:

Reviewer: "aumentar" -> "somar"

[source,@julia-setup]
----

> function somar_todos(t)

Reviewer: ```suggestion
function soma_todos(t)
```

end
----

> +total+ é iniciado em 0. Cada vez que passa pelo laço, +pass:[+=]+ captura um elemento da lista. O operador +pass:[+=]+ fornece um jeito fácil de atuaizar uma variável. Essa _declaração de atribuição aumentada_,

Reviewer: "Cada vez que passa pelo laço" -> "Toda vez que o laço é percorrido"

> +total+ é iniciado em 0. Toda vez que o laço é percorrido, +pass:[+=]+ captura um elemento da lista. O operador +pass:[+=]+ fornece um jeito fácil de atuaizar uma variável. Esse _comando de atribuição aumentada_,

Reviewer: ```suggestion
+total+ é iniciado em 0. Toda vez que o laço é percorrido, +pass:[+=]+ captura um elemento da lista. O operador +pass:[+=]+ fornece um jeito fácil de atualizar uma variável. Esse _comando de atribuição aumentada_,
```

(((comando de atribuição aumentada)))(((pass:[+=])))((("operador", "Base", "pass:[+=]", see="pass:[+=]")))

[source,julia]
----
total += x
----

é equivalente a

[source,julia]
----
total = total + x
----

Quando o laço é iniciado, +total+ acumula a soma dos elementos; uma variável usada dessa maneira é chamada de _acumuladora_.
(((acumulador)))

Adicionar elementos de uma lista é uma operação tão comum que o Julia fornece uma função interna, +sum+:

> (((soma)))((("função", "Base", "soma", see="soma")))

Reviewer: ```suggestion
(((sum)))((("função", "Base", "sum", see="sum")))
```

10
----

> Uma operação como essa que combina uma sequência de elementos a um único valor por vezes é chamado de _operação de redução_.

Reviewer: ```suggestion
Uma operação como essa que combina uma sequência de elementos a um único valor por vezes é chamada de _operação de redução_.
```

10
----

Uma operação como essa que combina uma sequência de elementos a um único valor por vezes é chamado de _operação de redução_.
(((operação de redução)))

> Muitas vezes, você deseja percorrer uma lista enquanto cria outra. Por exemplo, a função a seguir pega uma lista de strings e retorna uma nova string que contém strings de caracteres maiúsculas:

Reviewer: ".. uma nova string .. " -> " .. uma nova lista .. "

(((maiúscula)))((("função", "definido pelo programador", "maiúscula")))(((maiúscula)))

[source,@julia-setup]
----
functiontodas_maiúsculas(t)
    res = []
    for s in t
        push!(res, maiúscula(s))
    end
    res
end
----

> +res+ é inicializada com uma lista vazia; a cada vez no laço, anexamos o próximo elemento. Então, +res+ é outro tipo de acumulador.

Reviewer: "a cada vez no laço " -> "toda vez que o laço é percorrido"

> Um operador como +todas_maiúsculas+ é por vezes chamado de _mapa_ pois "mapeia" a função (neste caso +maiuscúla+) em cada um dos elementos em uma sequÊncia.

Reviewer: "(neste caso maiuscúla)" -> "(neste caso maiúscula)"
"sequÊncia" -> "sequência"

(((mapa)))

Outro tipo comum de operação é selecionar alguns dos elementos de uma lista e retornar uma sublista. Por exemplo, a seguinte função pega uma lista de strings e retorna uma lista que contém apenas strings maiúsculas:
(((apenas_maiusculas)))((("função", "definido pelo programador", "apenas maiusculas", see="apenas_maiusculas")))

[source,@julia-setup]
----
function apenas_maiusculas(t)
    res = []
    for s in t

>         if s == maiúscula(s)

Reviewer: aqui "uppercase" é uma função embutida do julia, então não podemos traduzir.

            push!(res, s)
        end
    end
    res
end
----

> Um operador como +apenas_maiusculas+ é chamado de _filtro_ pois seleciona alguns dos elementos e filtra em outro.

Reviewer: " .. e filtra em outro. " -> ".. e filtra outros. "

(((filtro)))

Operações mais comuns de listas podem ser exprressas como uma combinação de mapeamento, filtro e redução.

=== Sintaxe do Ponto

> Para cada operador binário como +pass:[^]+, existe uma correspondente em um _operador ponto_ pass:[<code>.^</code>] que é automaticamente definido para atuar +pass:[^]+ elemento-a-elemento nas listas. Por exemplo, pass:[<code>&#91;1, 2, 3&#93; ^ 3</code>] não é definida, mas pass:[<code>&#91;1, 2, 3&#93; .^ 3</code>] é definida computando o resultado elementar pass:[<code>&#91;1^3, 2^3, 3^3&#93;</code>]:

Reviewer: "existe uma correspondente em um _operador ponto_ " -> "existe um _operador ponto_ correspondente"

" ... não é definida .. " -> " .. não é definido .. "

" .. é definida computando o resultado elementar .. " -> " .. é definido como calcular elemento a elemento o resultado .. "

(((operador ponto)))((("operador", "base", ".", see="operador ponto")))(((".", see="operador ponto")))

[source,@julia-repl-test]
----
julia> print([1, 2, 3] .^ 3)
[1, 8, 27]
----

> Qualquer função +f+ do Julia pode ter aplicação do resultado elementar de qualquer lista com a _sintaxe do ponto_. Por exemplo, para deiar uma lista de stringas maiúsculas, não precisamos explicitar o laço:

Reviewer: " .. pode ter aplicação do resultado elementar ..  " -> " .. pode ser aplicada elemento a elemento à qualquer lista ... "

" .. para deiar uma lista de stringas maiúsculas .. " -> ".. para deixar uma lista de strings com todas as strings em letra maiúscula .. "

(((sintaxe do ponto)))

[source,@julia-repl-test]
----

> julia> t = maiúscula.(["abc", "def", "ghi"]);

Reviewer: "maiúscula" -> "uppercase"

julia> print(t)
["ABC", "DEF", "GHI"]
----

> Esse é um jeito elegante para criar mapeamentos. A função +todas_maiúsculas+ pode ser implementada de uma única linha:

Reviewer: " .. para criar mapeamentos .. " -> " .. de criar um mapeamento .. "
" ..de uma única linha .. " -> ".. em uma única linha .. "

(((todas_maiúsculas)))

[source,@julia-setup]
----
function todas_maiúsculas(t)

>     maiúscula.(t)

Reviewer: maiúscula -> uppercase

10
----

Uma operação como essa que combina uma sequência de elementos a um único valor por vezes é chamada de _operação de redução_.
(((operação de redução)))

> Muitas vezes, você deseja percorrer uma lista enquanto cria outra. Por exemplo, a função à seguir pega uma lista de strings e retorna uma nova lista que contém strings de caracteres maiúsculas:

Reviewer: ```suggestion
Muitas vezes, você deseja percorrer uma lista enquanto cria outra. Por exemplo, a função à seguir recebe uma lista de strings e retorna uma nova lista que contém as strings com todos os seus caracteres maiúsculos:
```

end
----

+res+ é inicializada com uma lista vazia; toda vez que laço é percorrido, anexamos o próximo elemento. Então, +res+ é outro tipo de acumulador.

> Um operador como +todas_maiúsculas+ é por vezes chamado de _mapa_ pois "mapeia" a função (neste caso +uppercase+) em cada um dos elementos em uma sequência.

Reviewer: ```suggestion
Uma operação como +todas_maiúsculas+ é por vezes chamado de _mapa_ pois "mapeia" uma função (neste caso +uppercase+) a cada um dos elementos em uma sequência.
```

(((mapa)))

> Outro tipo comum de operação é selecionar alguns dos elementos de uma lista e retornar uma sublista. Por exemplo, a seguinte função pega uma lista de strings e retorna uma lista que contém apenas strings maiúsculas:

Reviewer: ```suggestion
Outro tipo comum de operação é selecionar alguns dos elementos de uma lista e retornar uma sublista. Por exemplo, a seguinte função recebe uma lista de strings e retorna uma lista que contém apenas as strings com letras maiúsculas:
```

end
----

+res+ é inicializada com uma lista vazia; toda vez que laço é percorrido, anexamos o próximo elemento. Então, +res+ é outro tipo de acumulador.

> Uma operação como +todas_maiúsculas+ é por vezes chamado de _mapa_ pois "mapeia" uma função (neste caso +uppercase+) a cada um dos elementos em uma sequência.

Reviewer: ```suggestion
Uma operação como +todas_maiúsculas+ é por vezes chamado de _mapa_ pois “mapeia” uma função (neste caso +uppercase+) a cada um dos elementos em uma sequência.
```

(((mapa)))

Outro tipo comum de operação é selecionar alguns dos elementos de uma lista e retornar uma sublista. Por exemplo, a seguinte função recebe uma lista de strings e retorna uma lista que contém apenas as strings com letras maiúsculas:

> (((apenas_maiusculas)))((("função", "definido pelo programador", "apenas maiusculas", see="apenas_maiusculas")))

Reviewer: ```suggestion
(((apenas_maiúsculas)))((("função", "definido pelo programador", "apenas_maiúsculas", see="apenas_maiúsculas")))
```

[source,@julia-setup]
----

> function apenas_maiusculas(t)

Reviewer: ```suggestion
function apenas_maiúsculas(t)
```

end
----

> Um operador como +apenas_maiusculas+ é chamado de _filtro_ pois seleciona alguns dos elementos e filtra outros.

Reviewer: ```suggestion
Uma operação como +apenas_maiusculas+ é chamada de _filtro_ pois seleciona alguns dos elementos e filtra outros.
```

(((filtro)))

> Operações mais comuns de listas podem ser exprressas como uma combinação de mapeamento, filtro e redução.

Reviewer: ```suggestion
Operações mais comuns de listas podem ser expressas como uma combinação de mapeamento, filtro e redução.
```

=== Sintaxe do Ponto

> Para cada operador binário como +pass:[^]+, existe um _operador ponto_ pass:[<code>.^</code>] correspondente que é automaticamente definido para atuar +pass:[^]+ elemento-a-elemento nas listas. Por exemplo, pass:[<code>&#91;1, 2, 3&#93; ^ 3</code>] não é definido, mas pass:[<code>&#91;1, 2, 3&#93; .^ 3</code>] é definido como calcular elemento a elemento o resultado pass:[<code>&#91;1^3, 2^3, 3^3&#93;</code>]:

Reviewer: ```suggestion
Para cada operador binário como +pass:[^]+, existe um _operador ponto_ pass:[<code>.^</code>] correspondente que é automaticamente definido para efetuar +pass:[^]+ elemento-a-elemento nas listas. Por exemplo, pass:[<code>&#91;1, 2, 3&#93; ^ 3</code>] não é definido, mas pass:[<code>&#91;1, 2, 3&#93; .^ 3</code>] é definido como calcular elemento a elemento o resultado pass:[<code>&#91;1^3, 2^3, 3^3&#93;</code>]:
```

> Para cada operador binário como +pass:[^]+, existe um _operador ponto_ pass:[<code>.^</code>] correspondente que é automaticamente definido para efetuar +pass:[^]+ elemento-a-elemento nas listas. Por exemplo, pass:[<code>&#91;1, 2, 3&#93; ^ 3</code>] não é definido, mas pass:[<code>&#91;1, 2, 3&#93; .^ 3</code>] é definido como calcular elemento a elemento o resultado pass:[<code>&#91;1^3, 2^3, 3^3&#93;</code>]:

Reviewer: ```suggestion
Para cada operador binário como +pass:[^]+, existe um _operador ponto_ pass:[<code>.^</code>] correspondente que é automaticamente definido para efetuar +pass:[^]+ elemento-a-elemento em listas. Por exemplo, pass:[<code>&#91;1, 2, 3&#93; ^ 3</code>] não é definido, mas pass:[<code>&#91;1, 2, 3&#93; .^ 3</code>] é definido como calcular elemento a elemento o resultado pass:[<code>&#91;1^3, 2^3, 3^3&#93;</code>]:
```

> (((operador ponto)))((("operador", "base", ".", see="operador ponto")))(((".", see="operador ponto")))

Reviewer: ```suggestion
(((operador ponto)))((("operador", "Base", ".", see="operador ponto")))(((".", see="operador ponto")))
```

end
----

> Uma operação como +apenas_maiusculas+ é chamada de _filtro_ pois seleciona alguns dos elementos e filtra outros.

Reviewer: ```suggestion
Uma operação como +apenas_maiúsculas+ é chamada de _filtro_ pois seleciona alguns dos elementos e filtra outros.
```

['a', 'c']
----

> +splice!+ modifica a lista e retorna um elemento que foi removido.

Reviewer: ```suggestion
+splice!+ modifica a lista e retorna o elemento que foi removido.
```

+pop!+ deleta e retorna o último elemento:

> (((pop!)))((("função", "Base", "pop!", veja="pop!")))

Reviewer 2: ```suggestion
(((pop!)))((("função", "Base", "pop!", see="pop!")))
```

['a', 'b']
----

+popfirst!+ deleta e retorna o primeiro elemento:

> (((popfirst!)))((("funlão", "Base", "popfirst!", see="popfirst!")))

Reviewer: "funlão" -> "função"

['b', 'c']
----

> A função +pushfirst!+ e +push!+ insere um elemento no início e respectivamente no fim de uma lista.

Reviewer: "insere" -> "inserem" 

" no início e respectivamente no fim .. " -> "no início e no fim de uma lista, respectivamente, .. "

> (((pushfirst!)))((("função", "Base", "pushfirst", see="pushfirst!")))(((push!)))

Reviewer: ```suggestion
(((pushfirst!)))((("função", "Base", "pushfirst!", see="pushfirst!")))(((push!)))
```

> Se você não precisa remover valores, você pode usar a função +deleteat!+:

Reviewer: "Se você não precisa remover valores .. " -> " Se você não precisa do valor removido .. "

> (((deleteat!)))((("função", "Base", "delete!", see="deleteat!")))

Reviewer: "delete!" -> "deleteat!"

----

=== Listas e Strings

> Uma string é uma sequêncuia de caracteres e uma lista é uma sequência de valores, mas uma lista de caracteres não é o mesmo que uma string. Para converter uma string em uma lista de caracteres, voc~e pode usar a função +collect+:

Reviewer: "sequêncuia" -> "sequência"
"voc~e" -> "você"

> (((collect)))((("function", "Base", "collect", see="collect")))

Reviewer: "function" -> "função"

['s', 'p', 'a', 'm']
----

A função +collect+ divide uma sequência ou outra sequêcia em elementos individuais.

Se você quer dividir uma string em palavras, você pode usar a função +split+:
(((split)))((("função", "Base", "split", see="split")))

[source,@julia-repl-test]
----
julia> t = split("Vim lhe trazer este humilde presente");

julia> print(t)
SubString{String}["vim", "lhe", "trazer", "este", "humilde", "presente"]
----

> Um _argumento opcional_ chamado _delimitador_ especifica quais caracteres devem ser usados como limites de palavras. Os seguintes exemplos usan um hífen como um delimitador:

Reviewer: "usan" -> "usam"

(((argumento opcional)))(((delimitador)))

[source,@julia-repl-test]
----
julia> t = split("mayday-mayday-mayday", '-');

julia> print(t)
SubString{String}["mayday", "mayday", "mayday"]
----

> +join+ é o inverso de +split+. Ela pega uma lista de strings e concatena os elementos:

Reviewer: ```suggestion
+join+ é o inverso de +split+. Ela recebe uma lista de strings e concatena os elementos:
```

> (((joing)))((("função", "Base", "join", see="join")))

Reviewer: ```suggestion
(((join)))((("função", "Base", "join", see="join")))
```

[source,@julia-repl-test]
----
julia> t = ["Vim", "lhe", "trazer", "este", "humilde", "presente"];

julia> s = join(t, ' ')
"vim lhe trazer este humilde presente"
----

> No caso o delimitador é um caracter de espaço. Para concatenar strings sem espaços, você não precisa especificar um delimitador.

Reviewer: "caracter" -> "caractere"

> No caso o delimitador é um caractere de espaço. Para concatenar strings sem espaços, você não precisa especificar um delimitador.

Reviewer: ```suggestion
Neste caso o delimitador é um caractere de espaço. Para concatenar strings sem espaços, você não precisa especificar um delimitador.
```

=== Objetos e Valores

> Um _objeto_ é algo que uma variável pode se referir. Desde então, você poderia usar "objeto" e "valor" sem distinção.

Reviewer: ```suggestion
Um _objeto_ é algo que uma variável pode se referir a. Até agora, você poderia usar "objeto" e "valor" sem distinção.
```

> Um _objeto_ é algo que uma variável pode se referir a. Até agora, você poderia usar "objeto" e "valor" sem distinção.

Reviewer: ```suggestion
Um _objeto_ é algo que uma variável pode se referir a. Até agora, você poderia usar “objeto” e “valor” sem distinção.
```

> Um _objeto_ é algo que uma variável pode se refetir. À partir de agora, você pode usar "objeto" e "valor" intercambiavelmente.

Reviewer: " .. uma variável pode se refetir .. " -> " .. uma variável pode se referir a .. "

" À partir de agora, você pode usar "objeto" e "valor"  intercambiavelmente " -> " Desde então, você poderia usar "objeto" e "valor" sem distinção"

(((objeto)))(((variável)))(((valor)))

> Se você execusar estas declrações de atribuições:

Reviewer: "Se você execusar estas declrações de atribuições" -> "Se você executar estas declarações de atribuição"

[source,julia]
----
a = "banana"
b = "banana"
----

Sabemos que ambas +a+ e +b+ referem-se a uma string, mas não sabemos se eles referem à _mesma_ string. Existem dois estados possíveis, mostrados na Figura 10-2.
(((diagrama de estado)))

.Diagrama de estado
image::images/fig102.svg[]

> Em um caso, +a+ e +b+ referem-se a dois objetos diferentes que possuem o mesmo valor. No segundo caso, referem-se ao mesmo objeto.

Reviewer: ```suggestion
Em um caso, +a+ e +b+ referem-se a dois objetos diferentes que possuem o mesmo valor. No segundo caso, elas referem-se ao mesmo objeto.
```

> Para verificar se duas variáveis referem-se ao mesmo objeto, você pode usar o operador +=+ (*+\equiv TAB+*) ou +===+.

Reviewer 2: ```suggestion
Para verificar se duas variáveis referem-se ao mesmo objeto, você pode usar o operador +≡+ (*+\equiv TAB+*) ou +===+.
```

> Para verificar se duas variáveis se referem ao mesmo objeto, você pode usar +=+ (*+\equiv TAB+*)  ou o operador +===+.

Reviewer: referem -> referem-se

"você pode usar +=+ (*+\equiv TAB+*)  ou o operador +===+ " -> "você pode usar o operador +=+ (*+\equiv TAB+*) ou +===+. "

true
----

> Nesse exemplo, Julia criou apenas um objeto de sequência e ambos +a+ e +b+ se referem a ele. Mas quando você cria duas listas, você obtém dois objetos:

Reviewer: "... Julia criou apenas um objeto de sequência e ambos +a+ e +b+ se referem a ele " ->  " O Julia apenas criou um objeto string, e ambos +a+ e +b+ referem-se a ele "

false
----

> Portanto, o diagrama de estado se parece com <<<fig10-3>>.

Reviewer: ```suggestion
Portanto, o diagrama de estado se parece com <<fig10-3>>.
```

(((diagrama de estado)))

[[fig10-3]]
.Diagrama de estado
image::images/fig103.svg[]

> Nesse case poderíamos dizer que duas listas são _equivalentes_, porque possuem os mesmos elementos, mas não _indênticos_, porque eles não são os mesmos objetos. Se dois objetos são idênticos, eles também são equivalentes, mas se eles são equivalentes, eles não necessariamente são idênticos.

Reviewer: case -> caso

"porque eles não são os mesmos objetos " -> "porque ele não são o mesmo objeto"

> Nesse caso poderíamos dizer que as duas listas são _equivalentes_, porque possuem os mesmos elementos, mas não _indênticos_, porque elas não são o mesmo objeto. Se dois objetos são idênticos, eles também são equivalentes, mas se eles são equivalentes, eles não necessariamente são idênticos.

Reviewer: ```suggestion
Nesse caso poderíamos dizer que as duas listas são _equivalentes_, porque possuem os mesmos elementos, mas não _idênticos_, porque elas não são o mesmo objeto. Se dois objetos são idênticos, eles também são equivalentes, mas se eles são equivalentes, eles não necessariamente são idênticos.
```

(((equivalentes)))(((idênticos)))

> Para sermos mais precisos, um objeto possui um valor. Se você avaliar +[1, 2, 3]+, voc~e obterá uma lista de objetos cujos os valores são uma sequência de inteiros. Se uma outra lista possuir os mesmos elementos, dizemos que eles tem os mesmos valores, mas que não são os mesmos objetos.

Reviewer: voc~e -> você

"obterá uma lista de objetos cujos os valores são .. " -> "obterá um objeto lista cujo valor é uma .. "

" não são os mesmos objetos." -> "não são o mesmo objeto"

true
----

> O diagrama de estado se parece com <<<fig10-4>>.

Reviewer: ```suggestion
O diagrama de estado se parece com <<fig10-4>>.
```

[WARNING]
====
Embora esse comportamento possa ser útil, está propenso a erros. Em geral, é mais seguro evitar alias quando você estiver trabalhando com objetos mutáveis.
====

Para objetos imutáveis, como strings, o alias não é um problema. Neste exemplo:

[source,@julia-setup]
----
a = "banana"
b = "banana"
----

Quase nunca faz diferença se +a+ e +b+ referem-se a mesma string ou não.

=== Argumentos de Listas

Quando você passa uma lista para uma função, a função obtém uma referência para a lista. Se a função modifica a lista, o chamados vê a diferença. Por exemplo, +deletehead!+ remove o primeiro elemento de uma lista:

> (((deletehead!)))((("função", "definido pelo programador", "deletehead!", see="deletehead!")))(((popfirst)))

Reviewer: popfirst -> popfirst!

"o chamados vê a diferença " -> "quem a chama nota a diferença"

[WARNING]
====
Embora esse comportamento possa ser útil, está propenso a erros. Em geral, é mais seguro evitar alias quando você estiver trabalhando com objetos mutáveis.
====

Para objetos imutáveis, como strings, o alias não é um problema. Neste exemplo:

[source,@julia-setup]
----
a = "banana"
b = "banana"
----

Quase nunca faz diferença se +a+ e +b+ referem-se a mesma string ou não.

=== Argumentos de Listas

> Quando você passa uma lista para uma função, a função obtém uma referência para a lista. Se a função modifica a lista, quem chama nota a diferença. Por exemplo, +deleta_cabeça!+ remove o primeiro elemento de uma lista:

Reviewer: ```suggestion
Quando você passa uma lista para uma função, a função recebe uma referência para a lista. Se a função modifica a lista, quem chama nota a diferença. Por exemplo, +deleta_cabeça!+ remove o primeiro elemento de uma lista:
```

end
----

Segue abaixo como isto é utilizado:

[source,@julia-repl-test chap10]
----
julia> letras = ['a', 'b', 'c'];

julia> deletehead!(letras);

julia> print(letters)
['b', 'c']
----

O parâmetro +t+ e a variável +letras+ são alias para o mesmo objeto. O diagrama de estado se parece com <<fig10-5>>.
(((diagrama de estado)))

[[fig10-5]]
.Diagrama de estado
image::images/fig105.svg[]

> Desde que a lista seja compartilhada por dois quadros, eu desenhei entre eles.

Reviewer: "Desde que a lista seja compartilhada" -> "Já que a lista é compartilhada por dois quadros"

"eu desenhei entre eles" -> "eu a desenhei entre eles"

> Isto é importante para distinguir entre operações que modificam listas e operações que criam novas listas. Por exemplo, +push!+ modifica uma lista mas +vcat+ cria uma nova lista.

Reviewer: "Isto é importante para distinguir entre operações que modificam listas e operações que criam novas listas..." -> "É importante distinguir operações que modificam listas de operações que criam novas listas.."

" .. uma lista mas +vcat+ ..." -> " .. uma lista, mas +vcat+ ... " (vírgula)

[1, 2, 3, 4]
----

O resultado de +vcat+ é uma nova lista, e a lista original permanece inalterada.

Essa diferença é importante quando você escreve funções que devem modificar listas.

Por exemplo, essa função _não_ deleta a cabeça de uma lista:
(((baddeletehead)))((("função", "definido pelo programador", "baddeletehead", see="baddeletehead")))

[source,@julia-setup chap10]
----
function baddeletehead(t)
    t = t[2:end]                # EEERRROU!
end
----

> O operador de fatia cria uma nova lista e a atribuição faz com que +t+ se refira a ela, mas isso não afeta o chamador.

Reviewer: chamador -> quem a chama

[1, 2, 3, 4]
----

O resultado de +vcat+ é uma nova lista e a lista original permanece inalterada.

Essa diferença é importante quando você escreve funções que devem modificar listas.

Por exemplo, essa função _não_ deleta a cabeça de uma lista:
(((não_deleta_cabeça)))((("função", "definido pelo programador", "não_deleta_cabeça", see="não_deleta_cabeça")))

[source,@julia-setup chap10]
----
function não_deleta_cabeça(t)
    t = t[2:end]                # ERRADO!
end
----

O operador de fatia cria uma nova lista e a atribuição faz com que +t+ se refira a ela, mas isso não afeta quem a chama.
(((operador de fatia)))

[source,@julia-repl-test chap10]
----
julia> t4 = não_deleta_cabeça(t3);

julia> print(t3)
[1, 2, 3, 4]
julia> print(t4)
[2, 3, 4]
----

> No início de +não_deleta_cabeça+, +t+ e +t3+ se referem à mesma lista. No final, +t+ se refere a uma nova lista, mas +t3+ continua a se referir à original, uma lista não modificada.

Reviewer: ```suggestion
No início de +não_deleta_cabeça+, +t+ e +t3+ referem-se à mesma lista. No final, +t+ refere-se a uma nova lista, mas +t3+ continua a referir-se à lista original, que não foi modificada.
```

Uma alternativa é escrever uma função que cria e retorna uma nova lista. Por exemplo, +calda+ retorna todos, exceto o primeiro elemento de uma lista:

> (((tail)))((("função", "definido pelo programador", "calda", see="calda")))

Reviewer: ```suggestion
(((calda)))((("função", "definido pelo programador", "calda", see="calda")))
```

[source,@julia-setup chap10]
----
function calda(t)
    t[2:end]
end
----

Essa função deixa a lista original sem modificações. Veja como ela é usada:

[source,@julia-repl-test chap10]
----

> julia> calda = ['a', 'b', 'c'];

Reviewer 2: ```suggestion
julia> letras = ['a', 'b', 'c'];
```

julia> resto = calda(letras);

julia> print(rest)
['b', 'c']
----

=== Depurando

O uso descuidado de listas (e outros objetos mutáveis) pode levar longas horas para depurar. Veja algumas armadilhas comuns e maneiras de como evitá-las:
(((depurando)))

> * A maioria das funções de lista modifica o argumento. É o oposto das funções de sequência, que retornam uma nova sequência e deixam o original em paz.

Reviewer: ```suggestion
* A maioria das funções de lista modifica o argumento. É o oposto das funções de strings, que retornam uma nova string e deixam a original em paz.
```

+
Se você está acostumado a escrever código de string como:

> (((faixa)))(((sort!)))

Reviewer: ```suggestion
(((strip)))(((sort!)))
```

+
[source,julia]
----
nova_palavra = strip(palavra)
----
+
É tentador escrever código de lista assim:
+
[source,julia]
----
t2 = sort!(t1)
----
+
Como +sort!+ retorna a lista original modificada +t1+, +t2+ é um alias de +t1+.
+
[TIP]
====
Antes de usar funções e operadores de lista, você deve ler a documentação com cuidado e testá-los no modo interativo.
====

* Escolha um idioma e fique com ele.
+

> Parte do problema com listas é que existem muitas maneiras de fazer as coisas. Por exemplo, para remover um elemento de uma lista, você pode usar +pop!+, +popfirst!+, +Delete_at+ ou mesmo uma atribuição de fatia. Para adicionar um elemento, você pode usar +push!+, +pushfirst!+, +insert!+ Ou +vcat+. Supondo que +t+ é uma lista e +x+ é um elemento da lista, eles estão corretos:

Reviewer: ```suggestion
Parte do problema com listas é que existem muitas maneiras de fazer as coisas. Por exemplo, para remover um elemento de uma lista, você pode usar +pop!+, +popfirst!+, +delete_at+ ou até mesmo uma atribuição de fatia. Para adicionar um elemento, você pode usar +push!+, +pushfirst!+, +insert!+ Ou +vcat+. Supondo que +t+ é uma lista e +x+ é um elemento da lista, os seguintes estão corretos:
```

[2, 3, 4]
----

No início de +baddeletehead+, +t+ e +t3+ se referem à mesma lista. No final, +t+ se refere a uma nova lista, mas +t3+ continua a se referir à original, uma lista não modificada.

> Uma alternativa é escrever uma função que cria e retorna uma nova lista. Por exemplo, +rabo+ retorna todos, exceto o primeiro elemento de uma lista:

Reviewer: substituir "rabo" e "tail" por "calda" nos casos subsequentes.

end
----

> Essa função deixa a lista original sem modificações. Veja como é usado:

Reviewer: " como é usado" -> "como ela é usada"

[source,@julia-repl-test chap10]
----
julia> rabo = ['a', 'b', 'c'];

julia> resto = rabo(letras);

julia> print(rest)
['b', 'c']
----

=== Debugando

> O uso descuidado de listas (e outros objetos mutáveis) podem levar a longas horas de debug. Veja algumas armadilhas comuns e maneiras de como evitá-las:

Reviewer: "podem" -> "pode"
"de debug" -> "de depuração" ou "pode levar a muito tempo gasto debugando"

(((debugando)))

* A maioria das funções de lista modifica o argumento. É o oposto das funções de sequência, que retornam uma nova sequência e deixam o original em paz.
+
Se você está acostumado a escrever código de string como este:
(((faixa)))(((sort!)))
+
[source,julia]
----
new_word = strip(word)
----
+
É tentador escrever código de lista assim:
+
[source,julia]
----
t2 = sort!(t1)
----
+

> Como +sort!+ fetorna a lista original modificada +t1+, +t2+ é um alias de +t1+.

Reviewer: fetorna -> retorna

+
[TIP]
====
Antes de usar funções e operadores de lista, você deve ler a documentação com cuidado e testá-los no modo interativo.
====

* Escolha um idioma e fique com ele.
+

> Parte do problema com listas é que existem muitas maneiras de fazer as coisas. Por exemplo, para remover um elemento de uma lista, você pode usar +pop!+, +Popfirst!+, +Delete_at+ ou mesmo uma atribuição de fatia. Para adicionar um elemento, você pode usar +push!+, +Pushfirst!+, +Insert!+ Ou +vcat+. Supondo que +t+ é uma lista e +x+ é um elemento da lista, eles estão corretos:

Reviewer: tirar letra maiúscula dos nomes de função

append!(t, [x])
----
+
E os seguintes estão errados
+
[source,julia]
----

> insert!(t, 4, [x])         # ERROU!

Reviewer: ERROU -> ERRADO

push!(t, [x])              # ERROU!
----

* Faça cópias para evitar alias.
+

> Se você precisa usar uma função como +sort!+, que modifica o argumento, voc~e precisa manter a lista original, você pode fazer uma cópia:

Reviewer: voc~e -> você

> Se você precisa usar uma função como +sort!+, que modifica o argumento, você precisa manter a lista original, você pode fazer uma cópia:

Reviewer: ```suggestion
Se você quer usar uma função como +sort!+ que modifica o argumento, mas você precisa manter a lista original, você pode fazer uma cópia:
```

(((sort!)))
+
[source,@julia-repl-test chap10]
----
julia> t = [3, 1, 2];

> julia> t2 = t[:]; # t2 = cópia(t)

Reviewer: ```suggestion
julia> t2 = t[:]; # t2 = copy(t)
```

[1, 2, 3]
----
+

> Nesse exemplo, você também pode usar a função interna +sort+, que retorna uma nova lista classificada e deixa o original em paz:

Reviewer: classificada -> ordenada
"o original" -> "a original"

----

=== Glossário

lista (_array_)::
Uma sequência de valores.
(((lista)))

elemento::

> Um dos valores de uma lista (ou outra sequência), também chama items.

Reviewer: chama items -> chamado de itens

(((elemento)))

lista aninhada::

> Uma lista é um elemento de outra lista.

Reviewer: ```suggestion
Uma lista que é um elemento de uma outra lista.
```

(((lista aninhada)))

acumulador::
Uma variável que é utilizada em um laço para adicionar ou acumular resultados.
(((acumulador)))

atribuição aumentada::

> Uma atribuição que atualiza o valor de uma variável utilizando um operador +=+.

Reviewer: ```suggestion
Uma atribuição que atualiza o valor de uma variável utilizando um operador como +=+.
```

(((atribuição aumentada)))

operador ponto::
Um operador binário que é aplicado elemento a elemento de uma lista
(((operador ponto)))

sintaxe do ponto::

> Sintaxe utilizada para aplicar uma função resultado elementar a qualquer lista.

Reviewer: ```suggestion
Sintaxe utilizada para aplicar uma função elemento a elemento a qualquer lista.
```

(((sintaxe do ponto)))

operador de redução::
Um padrão de processamento que percorre uma sequência e acumula os elementos em um único resultado.
(((operador de redução)))

mapa::
Um padrão de processamento que percorre uma sequência e executa uma operação em cada elemento.
(((mapa)))

filtro::
Um padrão de processamento que percorre uma sequência e seleciona os elementos que atendem a algum critério.
(((filtro)))

objeto::
Algo que uma variável pode se referir. Um objeto tem um tipo e um valor.
(((objeto)))

equivalente::
Contém o mesmo valor.
(((equivalente)))

idêntico::

> Sendo o mesmo objeto (o que implica equivalencia).

Reviewer: ```suggestion
Ser o mesmo objeto (o que implica equivalência).
```

(((idêntico)))

referência::
Associação entre uma variável e seu valor.
(((referência)))

alias::

> Uma circunstância onde duas ou mais variáveis referem-se ao mesmo objeto.

Reviewer: ```suggestion
Uma circunstância no qual duas ou mais variáveis referem-se ao mesmo objeto.
```

(((alias)))

argumentos opcionais::
Argumentos que não são obrigatórios.
(((argumentos opcionais)))

delimitador::

> Um caracter ou ustring utilizada para indicar onde uma string deve ser cortada.

Reviewer: ustring -> string

> Um caracter ou string utilizada para indicar onde uma string deve ser cortada.

Reviewer: ```suggestion
Um caractere ou string utilizada para indicar onde uma string deve ser cortada.
```

(((delimitador)))

=== Exercícios

[source,@julia-eval chap10]
----
function soma_aninhada(t)
  total = 0

>   for nested in t

Reviewer: ```suggestion
  for aninhado in t
```

>     total += sum(nested)

Reviewer: ```suggestion
    total += sum(aninhado)
```

  nothing
end;

function é_ordenada(t)
  t == sort(t)
end;
----

[[ex10-1]]
===== Exercício 10-1

> Escreva uma função chamada +soma_aninhada+ que pega uma lista de listas com números inteiros e adicione os elementos de todas as listas aninhadas. Por exemplo:

Reviewer: ```suggestion
Escreva uma função chamada +soma_aninhada+ que recebe uma lista de listas de números inteiros e some os elementos de todas as listas aninhadas. Por exemplo:
```

(((soma_aninhada)))((("função", "definida pelo programador", "soma_aninhada", see="soma_aninhada")))

[source,@julia-repl-test chap10]
----
julia> t = [[1, 2], [3], [4, 5, 6]];

julia> soma_aninhada(t)
21
----

[[ex10-2]]
===== Exercício 10-2

> Escreva uma função chamada +soma_cumulativa+ que pega uma lista de números e retorne a soma cumulativa; isto é, uma nova lista em que o latexmath:[\(i\)]-ésimo elemento é a soma do primeiro elemento latexmath:[\(i\)] da lista original. Por exemplo:

Reviewer: ```suggestion
Escreva uma função chamada +soma_cumulativa+ que recebe uma lista de números e retorne a soma cumulativa; isto é, uma nova lista em que o latexmath:[\(i\)]-ésimo elemento é a soma do primeiro elemento latexmath:[\(i\)] da lista original. Por exemplo:
```

(((soma_cumulativa)))((("função", "definido pelo programador", "soma_cumulativa", see="soma_cumulativa")))

[source,@julia-repl-test chap10]
----
julia> t = [1, 2, 3];

julia> print(soma_cumulativa(t))
Any[1, 3, 6]
----

[[ex10-3]]
===== Exercício 10-3

> Escreva uma função chamada +interior+ que pega uma lista e retorna uma nova lista que contém tudo exceto o primeiro e o último elemento. Por exemplo:

Reviewer: ```suggestion
Escreva uma função chamada +interior+ que recebe uma lista e retorna uma nova lista que não contém o primeiro e o último elemento. Por exemplo:
```

----

[[ex10-4]]
===== Exercício 10-4

> Escreva uma função chamada +interior!+ que pega uma lista, modifique-a removendo o primeiro e o último elementos e retorne +nothing+. Por exemplo:

Reviewer: ```suggestion
Escreva uma função chamada +interior!+ que recebe uma lista, modifique-a removendo o primeiro e o último elemento e retorne +nothing+. Por exemplo:
```

----

[[ex10-5]]
===== Exercício 10-5

> Escreva uma função chamada +é_ordenada+ que use uma lista como parâmetro e retorne +true+ se a lista for classificada em ordem crescente e +false+ caso contrário. Por exemplo:

Reviewer: ```suggestion
Escreva uma função chamada +é_ordenada+ que use uma lista como parâmetro e retorne +true+ se a lista estiver ordenada em ordem crescente e +false+ caso contrário. Por exemplo:
```

(((é_ordenada)))((("função", "definido pelo programador", "é_ordenada", see="é_ordenada")))

[source,@julia-repl-test chap10]
----
julia> é_ordenada([1, 2, 2])
true
julia> é_ordenada(['b', 'a'])
false
----

[[ex10-6]]
===== Exercício 10-6

> Duas palavras são anagramas se você puder reorganizar as letras de uma para soletrar a outra. Escreva uma função chamada +é_anagrama+ que use duas strings e retorne + true + se forem anagramas.

Reviewer: ```suggestion
Duas palavras são anagramas se você puder reorganizar as letras de uma para formar a outra. Escreva uma função chamada +é_anagrama+ que recebe duas strings e retorne +true+ se elas forem anagramas.
```

(((é_anagrama)))((("função", "definida pelo programador", "é_anagrama", see="é_anagrama")))

[[ex10-7]]
===== Exercise 10-7

> Escreva uma função chamada +tem_repetidas+ que pegue uma lista e retorne +true+ se houver algum elemento que apareça mais de uma vez. Não deve modificar a lista original.

Reviewer: ```suggestion
Escreva uma função chamada +tem_duplicatas+ que recebe uma lista e retorne +true+ se houver algum elemento que apareça mais de uma vez. Ela não deve modificar a lista original.
```

> (((tem_repetidas)))((("função", "definida pelo programador", "tem_repetidas", see="tem_repetidas")))

Reviewer: ```suggestion
(((tem_duplicatas)))((("função", "definida pelo programador", "tem_duplicatas", see="tem_duplicatas")))
```

[[ex10-8]]
===== Exercício 10-8

Este exercício refere-se ao chamado Paradoxo de Aniversário, sobre o qual você pode ler em https://pt.wikipedia.org/wiki/Paradoxo_do_anivers%C3%A1rio.
(((Paradoxo de Aniversário)))

Se houver 23 alunos em sua turma, quais são as chances de vocês dois terem o mesmo aniversário? Você pode estimar essa probabilidade gerando amostras aleatórias de 23 aniversários e verificando correspondências.

> (((aleatório)))((("função", "Base", "aleatório", see="aleatório")))

Reviewer: ```suggestion
(((rand)))((("função", "Base", "rand", see="rand")))
```

[TIP]
====
Você pode gerar aniversários aleatórios com +rand(1:365)+.
====

[[ex10-9]]
===== Exercício 10-9

> Escreva uma função que leia o arquivo +words.txt+ e crie uma lista com um elemento por palavra. Escreva duas versões dessa função, uma usando +push!+ e a outra usando o idioma +t=[pass:[t ...], x]+. Qual deles demora mais para ser executado? Por quê?

Reviewer: ```suggestion
Escreva uma função que leia o arquivo +words.txt+ e crie uma lista com um elemento por palavra. Escreva duas versões dessa função, uma usando +push!+ e a outra usando o idioma +t=[pass:[t ...], x]+. Qual delas demora mais para ser executada? Por quê?
```

(((push!)))

[[ex10-10]]
===== Exercício 10-10

> Para verificar se uma palavra está na lista de palavras, você pode usar o operador +∈+, mas seria lento porque pesquisará as palavras em ordem.

Reviewer: ```suggestion
Para verificar se uma palavra está na lista de palavras, você poderia usar o operador +∈+, mas isto seria lento porque  as palavras seriam pesquisadas em ordem.
```

> Para verificar se uma palavra está na lista de palavras, você poderia usar o operador +∈+, mas isto seria lento porque  as palavras seriam pesquisadas em ordem.

Reviewer: ```suggestion
Para verificar se uma palavra está na lista de palavras, você poderia usar o operador +∈+, mas isto seria lento porque as palavras seriam pesquisadas em ordem.
```

> Como as palavras estão em ordem alfabética, podemos acelerar as coisas com uma pesquisa por bissecção (também conhecida como pesquisa binária), que é semelhante ao que você faz quando procura uma palavra no dicionário. Você começa no meio e verifica se a palavra que você procura vem antes da palavra no meio da lista. Nesse caso, você pesquisa a primeira metade da lista da mesma maneira. Caso contrário, você pesquisará a segunda metade.

Reviewer: ```suggestion
Como as palavras estão em ordem alfabética, podemos acelerar as coisas com uma busca bissecção (também conhecida como busca binária), que é semelhante ao que você faz quando procura uma palavra no dicionário. Você começa no meio e verifica se a palavra que você procura vem antes da palavra no meio da lista. Nesse caso, você pesquisa a primeira metade da lista da mesma maneira. Caso contrário, você pesquisará a segunda metade.
```

> De qualquer forma, você reduz pela metade o espaço restante da pesquisa. Se a lista de palavras tiver 113.809 palavras, serão necessárias 17 etapas para encontrar a palavra ou concluir que ela não existe.

Reviewer: ```suggestion
De qualquer forma, você reduz pela metade o espaço restante da busca. Se a lista de palavras possuir113.809 palavras, serão necessárias 17 etapas para encontrar a palavra ou concluir que ela não está na lista.
```

> De qualquer forma, você reduz pela metade o espaço restante da busca. Se a lista de palavras possuir113.809 palavras, serão necessárias 17 etapas para encontrar a palavra ou concluir que ela não está na lista.

Reviewer: ```suggestion
De qualquer forma, você reduz pela metade o espaço restante da busca. Se a lista de palavras possuir 113.809 palavras, serão necessárias 17 etapas para encontrar a palavra ou concluir que ela não está na lista.
```

> Escreva uma função chamada +é_bisseção+ que usa uma lista classificada e um valor-alvo e retorna +true+ se a palavra estiver na lista e +false+ se não estiver.

Reviewer: ```suggestion
Escreva uma função chamada +em_bisseção+ que usa uma lista ordenada e um valor-alvo e retorna +true+ se a palavra estiver na lista e +false+ se não estiver.
```

(((é_bisseção)))((("função", "definido pelo programador", "é_bisseção", see="é_bisseção")))

[[ex10-11]]
===== Exercício 10-11

> Duas palavras são um "par reverso" se cada uma for o inverso da outra. Escreva um programa +par_reverso+ que encontre todos os pares reversos na lista de palavras.

Reviewer: ```suggestion
Duas palavras são um “par reverso” se uma for o inverso da outra. Escreva um programa +par_reverso+ que encontre todos os pares reversos na lista de palavras.
```

Write a function called +inbisect+ that takes a sorted array and a target value and returns +true+ if the word is in the array and +false+ if it’s not.
(((inbisect)))((("function", "programmer-defined", "inbisect", see="inbisect")))
Escreva uma função chamada +em_bisseção+ que usa uma lista ordenada e um valor-alvo e retorna +true+ se a palavra estiver na lista e +false+ se não estiver.
(((é_bisseção)))((("função", "definido pelo programador", "é_bisseção", see="é_bisseção")))

[[ex10-11]]
===== Exercise 10-11
===== Exercício 10-11

Two words are a “reverse pair” if each is the reverse of the other. Write a program +reversepairs+ that finds all the reverse pairs in the word array.
(((reversepairs)))((("function", "programmer-defined", "reversepairs", see="reversepairs")))
Duas palavras são um “par reverso” se uma for o inverso da outra. Escreva um programa +par_reverso+ que encontre todos os pares reversos na lista de palavras.
(((par_reverso)))((("função", "definido pelo programador", "par_reverso", see="par_reverso")))

[[ex10-12]]

> ===== Exercise 10-12

Reviewer: isso também foi removido

> Duas palavras "interligam" se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, "micro" e "ondas" se interligam para formar "microondas".

Reviewer: Aqui acho que vale a pena encontrar outro exemplo, já que o exercício não se resume a simplesmente conectar duas palavras. A partir daqui as linhas também estão desalinhadas.

> Duas palavras "interligam" se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, "micro" e "ondas" se interligam para formar "microondas".

Reviewer: ```suggestion
Duas palavras “interligam” se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, “micro” e “ondas” se interligam para formar “microondas”.
```

Reviewer: inserir [[ex10-12]] na linha 1048

> . Escreva um programa que encontre todos os pares de palavras que se entrelaçam.

Reviewer: interligam ou entrelaçam?

> . Escreva um programa que encontre todos os pares de palavras que se entrelaçam.

Reviewer: ```suggestion
. Escreva um programa que encontre todos os pares de palavras que se interligam.
```

> . Você consegue encontrar alguma palavra intertravada de três vias; isto é, toda terceira letra forma uma palavra, começando na primeira, segunda ou terceira?

Reviewer: ```suggestion
. Você consegue encontrar alguma palavra interligada/entrelaçada de três vias; isto é, toda terceira letra forma uma palavra, começando na primeira, segunda ou terceira?
```

(((tem_duplicatas)))((("função", "definido pelo programador", "tem_duplicatas", see="tem_duplicatas")))

[[ex10-8]]
===== Exercício 10-8

Este exercício refere-se ao chamado Paradoxo de Aniversário, sobre o qual você pode ler em https://pt.wikipedia.org/wiki/Paradoxo_do_anivers%C3%A1rio.
(((Paradoxo de Aniversário)))

Se houver 23 alunos em sua turma, quais são as chances de vocês dois terem o mesmo aniversário? Você pode estimar essa probabilidade gerando amostras aleatórias de 23 aniversários e verificando correspondências.
(((rand)))((("função", "Base", "rand", see="rand")))

[TIP]
====
Você pode gerar aniversários aleatórios com +rand(1:365)+.
====

[[ex10-9]]
===== Exercício 10-9

Escreva uma função que leia o arquivo +words.txt+ e crie uma lista com um elemento por palavra. Escreva duas versões dessa função, uma usando +push!+ e a outra usando o idioma +t=[pass:[t ...], x]+. Qual delas demora mais para ser executada? Por quê?
(((push!)))

[[ex10-10]]
===== Exercício 10-10

Para verificar se uma palavra está na lista de palavras, você poderia usar o operador +∈+, mas isto seria lento porque as palavras seriam pesquisadas em ordem.

> Como as palavras estão em ordem alfabética, podemos acelerar as coisas com uma busca bissecção (também conhecida como busca binária), que é semelhante ao que você faz quando procura uma palavra no dicionário. Você começa no meio e verifica se a palavra que você procura vem antes da palavra no meio da lista. Nesse caso, você pesquisa a primeira metade da lista da mesma maneira. Caso contrário, você pesquisará a segunda metade.

Reviewer 2: ```suggestion
Como as palavras estão em ordem alfabética, podemos acelerar as coisas com uma busca por bissecção (também conhecida como busca binária), que é semelhante ao que você faz quando procura uma palavra no dicionário. Você começa no meio e verifica se a palavra que você procura vem antes da palavra no meio da lista. Nesse caso, você pesquisa a primeira metade da lista da mesma maneira. Caso contrário, você pesquisará a segunda metade.
```

De qualquer forma, você reduz pela metade o espaço restante da busca. Se a lista de palavras possuir 113.809 palavras, serão necessárias 17 etapas para encontrar a palavra ou concluir que ela não está na lista.

> Escreva uma função chamada +em_bisseção+ que usa uma lista ordenada e um valor-alvo e retorna +true+ se a palavra estiver na lista e +false+ se não estiver.
> (((é_bisseção)))((("função", "definido pelo programador", "é_bisseção", see="é_bisseção")))

Reviewer 2: ```suggestion
Escreva uma função chamada +em_bisseção+ que usa uma lista ordenada e um valor-alvo e retorna +true+ se a palavra estiver na lista e +false+ se não estiver.
(((em_bisseção)))((("função", "definido pelo programador", "em_bisseção", see="em_bisseção")))
```

[[ex10-11]]
===== Exercício 10-11

Duas palavras são um “par reverso” se uma for o inverso da outra. Escreva um programa +par_reverso+ que encontre todos os pares reversos na lista de palavras.
(((par_reverso)))((("função", "definido pelo programador", "par_reverso", see="par_reverso")))

[[ex10-12]]
===== Exercício 10-12

> Duas palavras "interligam" se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, "micro" e "ondas" se interligam para formar "microondas".

Reviewer: ```suggestion
Duas palavras “interligam” se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, “micro” e “ondas” se interligam para formar “microondas”.
```

> Duas palavras “interligam” se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, “micro” e “ondas” se interligam para formar “microondas”.

Reviewer 2: ```suggestion
Duas palavras “interligam” se receber letras alternadas de cada uma forma uma nova palavra. Por exemplo, “shoe” e “cold” se interligam para formar “schooled”.
```
É alternadamente, e vai ser utilizado a lista de palavras em inglês.
